doc: document self-referencing a package name (#31680)
nodejs/node: doc/api/esm.md

}
```

#### Self-referencing a package using its name

> Self-referencing a package's name in an `import` is allowed. For example,

Reviewer: ```suggestion
Within a package, the values defined in the package’s `package.json` `"exports"` field can be referenced via the package’s name. For example,
```

(and fix line length)

Author: Not sure about this change. It trades off simplicity for precision. Not sure the tradeoff is worth it in this case.

I will, of course, defer to your judgement.

Reviewer: I was confused that this only mentions `import`, when the feature applies to both `require` and `import`. I know you mention `require` further down, but I think we should lead off with that this feature is about `"exports"`, not `import`—as in, it works only when a package has `"exports"`, not when code uses `import`.

Author: Fixed

Author: Ah. Good point!

let's assume the `package.json` is:

```json
// package.json
{
  "name": "a-package",
  "exports": {
    ".": "./main.mjs",

>     "/cjs": "./main.js"

Reviewer 2: ```suggestion
    "./cjs": "./main.js"
```

Reviewer: I think having the one example for this involve two exports that are really the same code, just alternate ESM/CommonJS variants of it, confuses the purpose. Can we just leave dual-mode out of this, and make the exports be something like `".": "./main.mjs"` and "`./foo": "./foo.mjs"` or something?

Reviewer 2: I don't think that would work with the example below that uses `require`. We could make the import target  consistently CJS though.

Reviewer: I think consistently CommonJS is good. 👍

>     "./cjs": "./main.js"

Reviewer: ```js
  "exports": {
    ".": "./main.js",
    "./foo": "./foo.js"
```

I think my suggestion got swallowed up. Per conversation with @jkrems, I was suggesting we leave dual-mode out of this so as not to complicate the example. That means we need to use all-CommonJS sources so that the same `package.json` can work for a `require` example.

Author: It's not really dual mode, as it's not using conditional exports, but I see your point. I will change it to from "cjs" to "foo".

  }
}
```

Then any module _in that package_ can reference an export in the package itself:

```js
// ./a-module.mjs
import { something } from 'a-package'; // Imports "something" from ./main.mjs.
```

Self-referencing is available only if `package.json` has `exports`, and will

> allow importing only what that `export` allows. So the code below,

Reviewer: Do you mean “what `"exports"` allows”, or what the `export` keyword provides as a named export?

Author: You're right. It's confusing. I meant what the `exports` allows. I will make it plural, and add some more wording to make it more explicit.

Author: Fixed.

Reviewer: Always writing it as `"exports"` (with the quotes) can also make it clearer that it’s not a typo for `export`.

given the package above, will generate a runtime error:

```js
// ./another-module.mjs
import { another } from 'a-package/m.mjs'; // Imports "another" from ./m.mjs.
```

Self-referencing is also available when using `require`, both in an ES module,
and in a CommonJS one. For example, this code will also work:

```js
// ./a-module.js

> const { something } = require('a-package/cjs'); // Imports from ./main.js.

Reviewer 2: Maybe use "load" instead of "import", just in case?

```suggestion
const { something } = require('a-package/cjs'); // Loads from ./main.js.
```

}
```

#### Self-referencing a package using its name

Within a package, the values defined in the package’s
`package.json` `"exports"` field can be referenced via the package’s name.
For example, assuming the `package.json` is:

```json
// package.json
{
  "name": "a-package",
  "exports": {
    ".": "./main.mjs",
    "./foo": "./foo.js"
  }
}
```

Then any module _in that package_ can reference an export in the package itself:

```js
// ./a-module.mjs
import { something } from 'a-package'; // Imports "something" from ./main.mjs.
```

Self-referencing is available only if `package.json` has `exports`, and will
allow importing only what that `exports` (in the `package.json`) allows.
So the code below, given the package above, will generate a runtime error:

```js
// ./another-module.mjs

> import { another } from 'a-package/m.mjs'; // Imports "another" from ./m.mjs.

Reviewer: Can the comment say that this errors? Usually we include that kind of thing in the comments even when the main text also says it, because many people skip there main text and only read the code examples.

Author: Done.

```

Self-referencing is also available when using `require`, both in an ES module,
and in a CommonJS one. For example, this code will also work:

```js
// ./a-module.js

> const { something } = require('a-package/foo'); // Loads from ./main.js.

Reviewer: This should say foo.js to match the example.

Author: Are you sure? The `exports` field says `./foo` on the left side of the mapping.

Reviewer 2: ```suggestion
const { something } = require('a-package/foo'); // Loads from ./foo.js.
```

Author: Oh, oops. Looked at the wrong line. Thanks. Fixed.

```js
// ./another-module.mjs

// Imports "another" from ./m.mjs. Fails because
// the "package.json" "exports" field

> // does not allow this import.

Reviewer: ```suggestion
// does not provide an export named `m.mjs`.
```

“allow this import” makes me think that the problem is `{ another }`, not `m.mjs`.
